src/jsdoc_transformer.ts

```diff
@@ -112,88 +112,107 @@ export function maybeAddHeritageClauses(
-        mtt.debugWarn(decl, `could not resolve supertype: ${impl.getText()}`);
-        docTags.push({
-          tagName: '',
-          text: 'NOTE: tsickle could not resolve supertype, ' +
-              'class definition may be incomplete.\n'
-        });
+
+    // Otherwise, if we get here, we need to emit some jsdoc.
+    for (const expr of heritage.types) {
+      const parentName = heritageName(isExtends, expr);
+      // parentName may be null, indicating that the clause is something inexpressible
+      // in Closure, e.g. "class Foo implements Partial<Bar>".  For 'extends' clauses
+      // that means we cannot emit anything at all.
+      if (isExtends && !parentName) {
         continue;
       }
-      let alias: ts.Symbol = sym;
-      if (sym.flags & ts.SymbolFlags.TypeAlias) {
-        // It's implementing a type alias.  Follow the type alias back
-        // to the original symbol to check whether it's a type or a value.
-        const type = mtt.typeChecker.getDeclaredTypeOfSymbol(sym);
-        if (!type.symbol) {
-          // It's not clear when this can happen, but if it does all we
-          // do is fail to emit the @implements, which isn't so harmful.
-          continue;
-        }
-        alias = type.symbol;
-      }
-      if (alias.flags & ts.SymbolFlags.Alias) {
-        alias = mtt.typeChecker.getAliasedSymbol(alias);
+      docTags.push({
+        tagName: isExtends ? 'extends' : 'implements',
+        type: parentName || 'InexpressibleType',
+      });
+    }
+  }
+
+  /**
+   * Computes the Closure name of an expression occurring in a heritage clause,
+   * e.g. "implements FooBar".  Will return null if the expression is inexpressible
+   * in Closure semantics.  Note that we don't need to consider all possible
+   * combinations of types/values and extends/implements because our input is
+   * already verified to be valid TypeScript.  See test_files/class/ for the full
+   * cartesian product of test cases.
+   * @param isExtends True if were in an 'extends' false in an 'implements'.
+   */
+  function heritageName(isExtends: boolean, expr: ts.ExpressionWithTypeArguments): string|null {
+    const sym = mtt.typeChecker.getSymbolAtLocation(expr.expression);
+    if (!sym) {
+      // It's possible for a class declaration to extend an expression that
+      // does not have have a symbol, for example when a mixin function is
+      // used to build a base class, as in `declare MyClass extends
+      // MyMixin(MyBaseClass)`.
+      //
+      // Handling this correctly is tricky. Closure throws on this
+      // `extends <expression>` syntax (see
+      // https://github.com/google/closure-compiler/issues/2182). We would
+      // probably need to generate an intermediate class declaration and
+      // extend that.
+      mtt.debugWarn(decl, `could not resolve supertype: ${expr.getText()}`);
+      return null;
+    }
+
+    let alias: ts.Symbol = sym;
```

Comment:
my bad, but in the code below `alias` is kind of odd to read. Maybe you can find a way to write this such that below it's just `supertypeSymbol` or so?

Reply:
PTAL, I rearranged a bit (with no semantic changes so your presubmit should still be good)